webhooks/virtualmachine/validation/virtualmachine_validator_hardware_controllers.go

```diff
@@ -0,0 +1,203 @@
+// to controllers and that the number of devices per controller does not
+// exceed the maximum number of devices per controller.
+// We are not trying to be extensive here. The idea is to only perform
+// basic validations to the extent we can, and fall through to the
+// controller. It is possible that the VM is changed between the
+// webhook check and actual reconciliation, so this is best effort
+// anyway. We verify:
+//   - If a controllerBusNumber is specified, it must not be invalid
+//     (beyond max allowed).
+//   - If a controllerBusNumber is specified, there must be an entry in
+//     spec.hardware.controllers for this controller.
+//   - If a unitNumber is specified, it must not be invalid (beyond max allowed).
+//   - If a unitNumber is specified, it must not be reserved for the controller itself.
+//   - If a unitNumber is specified, it must not be already in use.
+func (v validator) validateControllerSlots(
+	_ *pkgctx.WebhookRequestContext,
+	vm, oldVM *vmopv1.VirtualMachine) field.ErrorList {
+
+	var allErrs field.ErrorList
+
+	volumes := vmopv1util.GetManagedVolumesWithPVC(*vm)
+	if len(volumes) == 0 {
+		return allErrs
+	}
+
+	volumesPath := field.NewPath("spec", "volumes")
+
+	var (
+		// specControllers builds a map of controller ID to controller
+		// specification from spec.hardware.controllers.
+		specControllers = vmopv1util.BuildVMControllersMap(vm)
+
+		// occupiedSlots builds a map of controller ID to a map of unit numbers
+		// that are already in use on the specified controller.
+		occupiedSlots = make(map[pkgutil.ControllerID]map[int32]struct{})
+	)
+
+	// Add CD-ROM controllers to the occupied slots to check for conflicts.
+	for _, cdrom := range vm.Spec.Hardware.Cdrom {
+		if cdrom.ControllerBusNumber != nil &&
+			cdrom.ControllerType != "" &&
+			cdrom.UnitNumber != nil && *cdrom.UnitNumber >= 0 {
+
+			controllerID := pkgutil.ControllerID{
+				ControllerType: cdrom.ControllerType,
+				BusNumber:      *cdrom.ControllerBusNumber,
+			}
+
+			if occupiedSlots[controllerID] == nil {
+				occupiedSlots[controllerID] = make(map[int32]struct{})
+			}
+			occupiedSlots[controllerID][*cdrom.UnitNumber] = struct{}{}
+		}
+	}
+
+	for i, vol := range vm.Spec.Volumes {
+		if vol.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc := vol.PersistentVolumeClaim
+		volPath := volumesPath.Index(i)
+
+		if pvc.ControllerBusNumber == nil ||
+			pvc.ControllerType == "" ||
+			pvc.UnitNumber == nil {
+			// These fields are validated by the virtualmachine validator's
+			// validateVolumes if they are required.
+			continue
+		}
+
+		controllerKey := pkgutil.ControllerID{
+			ControllerType: pvc.ControllerType,
+			BusNumber:      *pvc.ControllerBusNumber,
+		}
+
+		maxBusNumber := pvc.ControllerType.MaxPerVM()
+		minBusNumber := int32(0)
+		if oldVM == nil {
+			// If this is validating creating a VM, we reject bus number 0 because
+			// it is usually reserved for the default controller.
+			minBusNumber = int32(1)
```

Comment:
IC, so when creating the VM, even using that predefined controller is forbidden. Do you think it's ohk just to forbit creating or specifying bus 0 in spec.hardware? Using it is still ok? But then there is a chance that volume requires one type of controller while the default one has different. We want to avoid that kind of corner case I guess